### DISCUSSION TITLE
creazione fattura elettronica

### DISCUSSION THREAD

**Discussion by Alexxia5** [fattureincloud/api#428] [Q&A] [answered]
Buongiorno, sto cercando di creare una fattura elettronica ma mi torna il seguente errore:
y: {"error":{"message":"Invalid request.","validation_result":{"od_number":["The od number field is required when original document type is present."],"od_date":["The od date field is required when original document type is present."]}}}

ho settato i seguenti campi:
```
IssuedDocumentsApi apiInstance = new IssuedDocumentsApi(defaultClient);
        Client cliente = dettaglioFattura.getClienteDaFatturare();

        Entity entity = new Entity()
                .id(cliente.getId()) // Only if the client already exists
                .name(dettaglioFattura.getClienteDaFatturare().getName())
                .vatNumber(dettaglioFattura.getClienteDaFatturare().getVatNumber())
                .taxCode(dettaglioFattura.getClienteDaFatturare().getTaxCode())
                .addressStreet(dettaglioFattura.getClienteDaFatturare().getAddressStreet())
                .addressPostalCode(dettaglioFattura.getClienteDaFatturare().getAddressPostalCode())
                .addressCity(dettaglioFattura.getClienteDaFatturare().getAddressCity())
                .addressProvince(dettaglioFattura.getClienteDaFatturare().getAddressProvince())
                .country(dettaglioFattura.getClienteDaFatturare().getCountry())
                .email(dettaglioFattura.getClienteDaFatturare().getEmail())
                ;
        IssuedDocument invoice = new IssuedDocument()
                .type(IssuedDocumentType.INVOICE)
                .entity(entity)
                .date(dettaglioFattura.getDataFattura())
              //  .number(2024000545) // se si omette il numero verrà incrementato automaticamente
                .numeration(numeration) // inserito solo per diversificare l'ambiente di collaudo e demo
                .subject(dettaglioFattura.getSubject()) // non visibile in fattura
                .visibleSubject(dettaglioFattura.getVisibleSubject()) // visibile in fattura
                .currency(new Currency().id("EUR"))
                .language(new Language()
                        .code("it")
                        .name("italiano")
                ).notes(VISITA_PRELIM+" "+dettaglioFattura.getNomePalestra()+"<br><br>"+PAGAMENTO_CON+" "+dettaglioFattura.getPagamentoCon())
                .addItemsListItem(
                        new IssuedDocumentItemsListItem()
                                .productId(dettaglioFattura.getVociDaFAtturare().get(0).getId())
                                .code(dettaglioFattura.getVociDaFAtturare().get(0).getCode())
                                .name(dettaglioFattura.getVociDaFAtturare().get(0).getName())
                                .netPrice(dettaglioFattura.getImportoBigDecimal())
                                .discount(BigDecimal.valueOf(0))
                                .qty(BigDecimal.valueOf(1))
                                .vat(new VatType().id(TLMDConstants.FIC_IVA_O_ART_10_COMMA1))  // iva al 22% id=0  // 0% id=16 // 4%  id=4
                                .description(DOTT+" "+dettaglioFattura.getNomeMedico())
                ).addPaymentsListItem(
                        new IssuedDocumentPaymentsListItem()                               
                                .dueDate(dettaglioFattura.getDataPagamento()) //scadenze
                                .paidDate(dettaglioFattura.getDataPagamento())
                                .status(IssuedDocumentStatus.PAID)
                                .paymentAccount(new PaymentAccount().id(dettaglioFattura.getPaymentAccount()))
                )
                .eInvoice(true)
                .eiData(
                        new IssuedDocumentEiData()
                                .paymentMethod("MP05")
                )
CreateIssuedDocumentRequest createIssuedDocumentRequest = new CreateIssuedDocumentRequest()
                .data(invoice)
                .options(
                        new IssuedDocumentOptions()
                                .fixPayments(true)
                );
```

come se volesse l'id di un documento già presente. 
MI potete aiutare a capire.
Grazie

**Comment by emajo** [comment]
Ciao @Alexxia5 ti consiglio di scrivere al nostro supporto clienti per andare a vedere i log nel dettaglio di ciò che sta arrivando alle API. Sarebbe utile sapere anche la versione della SDK che sta usando, e, in caso non fosse l'ultima consiglio di aggiornarla e ricontrollare se il problema persiste.

Grazie,
Emanuele.

**Reply by valmoz** [reply]
Buongiorno @Alexxia5 
credo di essere riuscito a risalire al log della sua chiamata, anche se non sono sicuro non avendo a disposizione il suo company id (per quello deve passare dal customer support).
Dal suo JSON vedo che sta settando il campo original_document_type a "ordine", come indicato anche dall'errore di validazione:

```
"ei_data": {
      "original_document_type": "ordine",
      "payment_method": "MP05"
    },
```

Ciò non corrisponde al codice che ci ha condiviso, quindi le chiederei di ricontrollare il codice e di utilizzare l'ultima versione dell'SDK (2.0.21).

Può anche controllare il codice generato come spiegato qui: https://developers.fattureincloud.it/docs/sdks/java-sdk#spiral_notepad-retrieve-the-json-request-body

In alternativa può contattare il nostro supporto clienti, allegando questa discussione, in modo tale da poter ottenere il suo company id e controllare dai log delle API. Poiché il supporto clienti utilizza chat private, è anche possibile condividere il JSON originale.

Grazie
Mauro

**Comment by Alexxia5** [comment]
Ciao, intanto grazie per la risposta.
La versione che sto utilizzando è la 2.0.20, va bene?
Ho controllato, il codice postato è esattamente quello.
Il company id è 1305751. Sto utilizzando un account di prova per i test.
Puoi dirmi cosa deve contenere il campo "original_document_type"? 
Io devo creare una fattura elettronica che non è legata a nessun'altro tipo di documento. 
Riesco tranquillamente a creare fatture in pdf, ma non so come creare (o convertirle) in elettroniche.
Pensavo che bastasse aggiungere la parte 
  .eInvoice(true)
  .eiData(
      new IssuedDocumentEiData()
         .paymentMethod("MP05")
  )

ma non va bene, mi ritorna l'errore
y: {"error":{"message":"Invalid request.","validation_result":{"od_number":["The od number field is required when original document type is present."],"od_date":["The od date field is required when original document type is present."]}}}

Ma non c'è un documento originale su cui legare la fattura. 
cosa deve contenere il campo od_number e od_Date?

**Reply by valmoz** [reply]
Buongiorno @Alexxia5 
il campo original_document_type dovrebbe essere omesso per la sua casistica, ma dai log che ho trovato (e non sono sicuro al 100% che siano suoi, non avendo a disposizione il Company ID) e dall'errore che viene restituito si evince che invece la sua richiesta contiene un original_document_type impostato su "ordine".
Le consiglio di provare a ottenere il JSON come le ho indicato nel precedente messaggio, per verificare perché viene impostato tale valore.

In alternativa provi a scrivere a info@fattureincloud.it, allegando questa discussione, in modo tale che il supporto clienti possa recuperare il suo Company ID.

Grazie
Mauro

**Reply by valmoz** [reply]
scusi, ho visto ora che ha inserito il company ID nella sua precedente risposta.
le confermo che sta popolando il campo original_document_type
```
"ei_data": {
      "original_document_type": "ordine",
      "payment_method": "MP08"
    },
```
Dovrebbe procedere a controllare il JSON come indicato in precedenza.

Grazie
Mauro

**Reply by valmoz** [reply]
Buongiorno @Alexxia5 
abbiamo effettuato un controllo, ed effettivamente sembra che le versioni 2.0.20 e 2.0.21 della Java SDK abbiano un bug che va ad impostare il default per il parametro indicato.
Appena possibile procederemo a rilasciare una nuova versione corretta dell'SDK.

Grazie
Mauro

**Reply by Alexxia5** [reply]
perfetto.Grazie mille
Nel frattempo per testare la creazione della fattura elettronica provo a creare un ordine e a legarla a quello tanto per vedere se viene creata o se ci sono altri problemi.
Un'altra domanda.
E' possibile creare una fattura pdf tramite Api e successivamente convertirla, direttamente dal sito fatture in cloud, in fattura elettronica?
Oppure bisogna generare direttamente con le Api la fattura elettronica?
Grazie ancora

**Reply by Alexxia5** [reply]
Sono riuscita a creare la fattura elettronica legata ad un ordine fittizio che poi eliminerò quando rilascerete la nuova versione in cui non verrà inserito di default    "original_document_type": "ordine".

Nell'attesa delle versione aggiornata vi ringrazio e saluto
Alessandra

**Reply by valmoz** [reply, marked as answer]
buongiorno @Alexxia5 
quando ha altre domande legate ad un altro argomento le suggerisco di aprire una nuova discussione dedicata, in modo tale che sia più semplice per gli altri utenti cercare eventuali risposte preesistenti.

Grazie
Mauro

**Reply by emajo** [reply]
Ciao @Alexxia5, abbiamo rilasciato la nuova versione dell'SDK (2.0.22) contenente il fix per il problema da lei riportato.

Grazie,
Emanuele.

**Reply by Alexxia5** [reply]
@valmoz va bene.
@emajo  . Importerò la nuova versione. Vi farò sapere.
Grazie mille
Alex

**Reply by Alexxia5** [reply]
Risolto.
Potete chiudere il post. 
Grazie mille.
